active/0021-transfer-files-over-mqtt.md

```diff
@@ -0,0 +1,205 @@
+# MQTT based file transfer
+
+## Changelog
+
+* 2022-09-20: @qzhuyan, @id Initial draft
+
+## Abstract
+
+TODO:
+
+## Motivation
+
+EMQX customers are asking for file transfer functionality from IoT devices to their cloud (primary use case), and from cloud to IoT devices over MQTT. Right now they are uploading files from devices via FTP or HTTPS (e.g. to S3), but this approach has downsides:
+
+* FTP and HTTP servers usually struggle to keep up with large number of simultaneous bandwidth-intensive connections
+* packet loss or reconnect forces clients to restart the transfer
+* devices which already talk MQTT need to integrate with one more SDK, address authenticaion and authorization, and potentially go through an additional round of security audit
+
+Known cases of device-to-cloud file transfer:
+
+* [CAN bus](https://en.wikipedia.org/wiki/CAN_bus) data
+* Image taken by industry camera for Quality Assurance
+* Large data file collected from forklift
+* Video and audio data from truck cars, and video data captured by inbound unloading cameras
+* Vehicle real-time logging, telemetry, messaging
+* Upload collected ML logs
+
+Known cases of cloud-to-device file transfer:
+
+* Upload AI/ML models
+* Firmware upgrades
+
+## Terminology
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in [BCP 14](https://www.rfc-editor.org/bcp/bcp14) [[RFC2119](https://www.rfc-editor.org/rfc/rfc2119)] [[RFC8174](https://www.rfc-editor.org/rfc/rfc8174)] when, and only when, they appear in all capitals, as shown here.
+
+The following terms are used as described in [MQTT Version 5.0 Specification](https://docs.oasis-open.org/mqtt/mqtt/v5.0/os/mqtt-v5.0-os.html#_Toc39010030):
+* Application Message
+* Server
+* Client
+* Topic Name
+* Topic Filter
+* MQTT Control Packet
+
+*At least once*: a message can be delivered many times, but cannot be lost
+
+## Requirements
+
+* The protocol MUST use only PUBLISH type of MQTT Control Packet
+* The protocol MUST support transfer of file segments
+* Server MUST be able to verify integrity of each file segment
```

Comment:
Is there a way to verify the integrity after putting all segments together?
Everything ends, right? even if it's a stream.

Reply:
final sha256 as part of the `fin` message? wdyt?